### PR TITLE
Adding ChangeInstrument op

### PR DESCRIPTION
**Pull request by wan9c9**
## Description

The ChangeInstrument allows using the features of another instrument.

## How Has This Been Tested?

- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [x] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [ ] Fix bugs
- [x] Add new feature
- [ ] Update documentation

### CODE REVIEW

**Comment by you-n-g** [comment]
@wan9c9 

Please fix [the CI](https://github.com/microsoft/qlib/runs/5659318871?check_suite_focus=true#step:4:45)

File: qlib/utils/__init__.py [hunk 1]
```diff
@@ -280,6 +280,9 @@ def parse_field(field):
         (r"(\w+\s*)\(", r"Operators.\1("),
     ]:  # Features  # Operators
         field = re.sub(pattern, new, field)
+    #for ChangeInstrument
```
Comment (you-n-g) [inline qlib/utils/__init__.py:283]: I don't think this substitution.
If you don't add this lines of code. I think the expression below will work
"ChangeInstrument('SH600519',$close)"

File: qlib/data/ops.py [hunk 2]
```diff
@@ -1541,6 +1626,7 @@ def _load_internal(self, instrument, start_index, end_index, *args):
 
 TOpsList = [TResample]
 OpsList = [
+    ChangeInstrument,
```
Comment (you-n-g) [inline qlib/data/ops.py:1629]: I'm not sure if I'm correct.
Personally, it confuses me a little when I see the word change.
Do you think "FixedInst" will be a better name?

File: qlib/data/ops.py [hunk 3]
```diff
@@ -32,6 +32,91 @@
...
+    """
+
+    def __init__(self, instrument, feature):
+        self.instrument = instrument
+        self.feature = feature
+
+    def __str__(self):
+        return "{}({},{})".format(type(self).__name__, self.instrument, self.feature)
+
+    def load(self, instrument, start_index, end_index, freq):
```
Comment (you-n-g) [inline qlib/data/ops.py:69]: `super().load(self, instrument, start_index, end_index, *args)`
Will it better to call the method from super class directly?

File: qlib/data/ops.py [hunk 4]
```diff
@@ -32,6 +32,91 @@
 
 np.seterr(invalid="ignore")
 
+
+#################### Change instrument ########################
```
Comment (you-n-g) [inline qlib/data/ops.py:36]: This duplicates with the docs string.
I think we can keep only the one in the docstring.

**Comment by you-n-g** [comment]
Hi,  @wan9c9 
Your implementation is really smart! I think it will benefit a lot of users.
I add several comments just now. It will be great if you can refine it :)

File: qlib/data/ops.py (hunk 4 above)
Comment (wan9c9) [reply qlib/data/ops.py:36]: Yes, sure. I wrote the first paragraph to motivate the class, and it should be fine to keep only those in the doc string and I'll update that.

File: qlib/data/ops.py (hunk 2 above)
Comment (wan9c9) [reply qlib/data/ops.py:1629]: Maybe SwitchInstrument or ReplaceInstrument is a better name? Although it's unlikely to have multiple chained ChangeInstrument op, but in principle one can do that. FixedInst implies one can't do that.

File: qlib/utils/__init__.py (hunk 1 above)
Comment (wan9c9) [reply qlib/utils/__init__.py:283]: Without the two lines below, I would see the following error for the following code: 
marketRet = ChangeInstrument('SH000300', Feature('close')/Ref(Feature('close'),1) -1)
and use QlibDataLoader to load the values.

expression = eval(parse_field(field))
  File "<string>", line 1, in <module>
NameError: name 'SH000300' is not defined.

It's likely related to how quotes are dealt with in the sub function, but I didn't dive into it. With the two lines below, the error was gone. If you know the exact reason, please let me know. Thanks!

File: qlib/data/ops.py (hunk 3 above)
Comment (wan9c9) [reply qlib/data/ops.py:69]: Please let me know how you think and I can revise it. Thanks!

File: qlib/utils/__init__.py (hunk 1 above)
Comment (you-n-g) [reply qlib/utils/__init__.py:283]: I remove the two lines and implemented a test. It passes.
Could you give a examples or test about how to reproduce your error ?
Thanks

File: qlib/data/ops.py (hunk 2 above)
Comment (you-n-g) [reply qlib/data/ops.py:1629]: OK.
I think `ChangeInstrument` will be OK now.

File: qlib/data/ops.py (hunk 3 above)
Comment (you-n-g) [reply qlib/data/ops.py:69]: Yes. 
I think this is a more concise way  to implement.

> Another way to achieve the same caching performance may be to call super().load(self, self.instrument, start_index, end_index, *args). This should do the trick without changing the _load_internal function, right?

**Comment by you-n-g** [comment]
Hi, @wan9c9 
I've replied to your questions about the issues.
Looking forward to your reply :).

BTW,  please fix the error reported by CI

File: qlib/data/ops.py (hunk 3 above)
Comment (you-n-g) [reply qlib/data/ops.py:69]: @wan9c9
Hi, did you encounter any problems when implementing the concise version  ?

File: qlib/data/ops.py (hunk 4 above)
Comment (you-n-g) [reply qlib/data/ops.py:36]: Hi, @wan9c9  did you fix this ? 
Thanks

**Comment by you-n-g** [comment]
@wan9c9  did you encounter any problems when fix the remained issues in the comments?

**Comment by wan9c9** [comment]
@you-n-g, sorry for the delays, I'm busy with other stuff and couldn't commit the revision as I also messed up the code. I did wrote some tests, which couldn't pass the test. The test code are show below. 




class TestOperatorDataSetting(TestOperatorData):
    def test_setting(self):

        def test_case(instruments, queries, note=None):
            if note:
                print(note)
            print(f"checking {instruments} with queries {queries}")
            df = D.features(instruments, queries)
            print(df)
            return df

        test_case(["SH600519"], ["ChangeInstrument('SH000300', $close)"], "get market index close")
        test_case(["SH600519"], ["ChangeInstrument('SH000300', Feature('close')/Ref(Feature('close'),1) -1)"],
                  "get market index return with Feature")
        test_case(["SH600519"], ["ChangeInstrument('SH000300', $close/Ref($close,1) -1)"],
                  "get market index return with expression")
        test_case(["SH600519"], ["($close/Ref($close,1) -1) - ChangeInstrument('SH000300', $close/Ref($close,1) -1)"],
                  "get excess return with expression with beta=1")

        ret = "Feature('close') / Ref(Feature('close'), 1) - 1"
        benchmark = 'SH000300'
        n_period = 252
        marketRet = f"ChangeInstrument(\'{benchmark}\', Feature('close') / Ref(Feature('close'), 1) - 1)"
        marketVar = f"ChangeInstrument(\'{benchmark}\', Var({marketRet}, {n_period}))"
        beta = f"Cov({ret}, {marketRet}, {n_period}) / {marketVar}"
        excess_return = f"{ret} - {beta}*({marketRet})"
        fields = ["Feature('close')", f"ChangeInstrument(\'{benchmark}\', Feature('close'))", ret, marketRet, beta, excess_return]
        test_case(["SH600519"],fields[5:],
                  "get market beta and excess_return with estimated beta")

        # All the queries above passed

        """
        the code below did not pass if  the following lines are not added to qlib/utils/__init__.py
        but will pass if added.
        # for ChangeInstrument
        field = re.sub(r"Operators.ChangeInstrument\((\w+),",
                       rf'Operators.ChangeInstrument("\1",', field)
        """

        from qlib.data.dataset.loader import QlibDataLoader
        from qlib.data.ops import Feature, Ref, ChangeInstrument, Var, Cov
        instrument = "SH600519"
        ret = Feature('close') / Ref(Feature('close'), 1) - 1
        benchmark = 'SH000300'
        n_period = 252
        marketRet = ChangeInstrument(benchmark, Feature('close') / Ref(Feature('close'), 1) - 1)
        marketVar = ChangeInstrument(benchmark, Var(marketRet, n_period))
        beta = Cov(ret, marketRet, n_period) / marketVar
        fields = [Feature('close'), ChangeInstrument(benchmark, Feature('close')),
                  ret, marketRet,
                  beta, ret - beta*marketRet]
        names = ['close', 'marketClose', 'ret', 'marketRet', f"beta_{n_period}","excess_return"]
        data_loader_config = {"feature": (fields,
                                          names
                                          )
                              }
        data_loader = QlibDataLoader(config=data_loader_config)
        df = data_loader.load(instruments=[instrument])#, start_time=start_time)
        print(df)

        #test_case(["SH600519"],fields,
                  #"get market beta and excess_return with estimated beta")

if __name__ == "__main__":
    unittest.main()


Could you please take a look? 

Thanks!

**Comment by you-n-g** [comment]
@wan9c9 I added your test case here and fix a bug. Please merge it
https://github.com/wan9c9/qlib/pull/4wan9c9
Is the problem solved? Do you have other issues when fixing the remain comments in this PR?

**Comment by wan9c9** [comment]
@you-n-g, looks like the link https://github.com/wan9c9/qlib/pull/4wan9c9 doesn't work. Can you please send a valid link and I'll double check?  Thanks!

**Comment by you-n-g** [comment]
> @you-n-g, looks like the link https://github.com/wan9c9/qlib/pull/4wan9c9 doesn't work. Can you please send a valid link and I'll double check? Thanks!

Sorry I mixed your ID and the link.   Here is  the right link https://github.com/wan9c9/qlib/pull/4

**Comment by you-n-g** [comment]
@wan9c9  It looks great now!
Thanks for your contribution!

**Pull request merged**